aea/protocols/dialogue/base.py

```diff
@@ -614,7 +609,14 @@ def _has_message_id(self, message_id: int) -> bool:
         if self.is_empty:
             return False
 
-        return self.STARTING_MESSAGE_ID <= message_id <= self.last_message.message_id  # type: ignore
+        if abs(message_id) < self.STARTING_MESSAGE_ID:
+            return False
+
+        for msg in chain(self._incoming_messages, self._outgoing_messages):
+            if msg.message_id == message_id:
+                return True
+
+        return False
```

Comment:
Instead of looping through all messages, you could simply check if it is between 0 and the last negative/positive message id, if message_id itself is negative/positive respectively.

Reply:
Here we can indeed use the fact that the message id on either side needs to be monotonically increasing/decreasing.

Reply:
done